File: java/dirty/src/test/java/fr/cleancode/BlockChainTest.java
package fr.cleancode;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.fail;

import org.junit.Before;
import org.junit.Test;

public class BlockChainTest {
    private BlockChain blockchain;

    private Block add(String parentHash, String data) {
        final Block block = new Block(parentHash, data);
        blockchain.addBlock(block);
        return block;
    }

    @Before
    public void createBlockChain() {
        blockchain = new BlockChain();
    }

    @Test
    public void testInitialBlock() {
        assertThat(blockchain.blocks.keySet())
                .contains("733ad20af52ec1902f51e97b99444754df7db2bf91bb26e076723c5e9fc12bc1");
    }

    @Test
    public void testAddRootBlock() {
        Block rootBlock = new Block(null, "2nd root block");
        boolean added = blockchain.addBlock(rootBlock);

        assertThat(added).isFalse();
    }

    @Test
    public void testAddBlock() {
        // nominal case for addBlock
        //TODO
        fail("TODO");
    }

    @Test
    public void testAddBlockWrongParent() {
        // adding a block with a parent not in the block chain
        //TODO
        fail("TODO");
    }

    @Test
    public void testGetLastBlock() {
        Block root = blockchain.blocks.get("733ad20af52ec1902f51e97b99444754df7db2bf91bb26e076723c5e9fc12bc1");
        Block last = blockchain.getLastBlock();

        assertThat(last).isSameAs(root);
    }

    /*
     * Given this block chain:
     * a--b--c--d--e--f--g
     *     \-h--i
     * After adding a new block 'j' we should get:
     * a--b--c--d--e--f--g--j
     */
    @Test
    public void testCleanBlockChain() {
        Block last = blockchain.getLastBlock();
        Block a = add(last.hash, "a");
        Block b = add(a.hash, "b");
        Block c = add(b.hash, "c");
        Block d = add(c.hash, "d");
        Block e = add(d.hash, "e");
        Block f = add(e.hash, "f");
        Block g = add(f.hash, "g");
        Block h = add(b.hash, "h");
        Block i = add(h.hash, "i");

        Block j = add(g.hash, "j");

        assertThat(blockchain.blocks.keySet()).doesNotContain(h.hash);
        assertThat(blockchain.blocks.keySet()).doesNotContain(i.hash);
        assertThat(blockchain.blocks.keySet()).contains(j.hash);
    }

}

File: csharp/dirty/fr.soat.cleancode.tests/BlockChainTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace fr.soat.cleancode.tests
{
    [TestClass]
    public class BlockChainTest
    {

        private BlockChain blockChain;

        private Block Add(string parentHash, string data)
        {
            var block = new Block(parentHash, data);
            blockChain.AddBlock(block);
            return block;
        }

        [TestInitialize]
        public void CreateBlockChain()
        {
            blockChain = new BlockChain();

        }

        [TestMethod]
        public void TestInitialBlock()
        {
            CollectionAssert.Contains(blockChain.blocks.Keys, "733ad20af52ec1902f51e97b99444754df7db2bf91bb26e076723c5e9fc12bc1");
        }


        [TestMethod]
        public void TestAddRootBlock()
        {
            var rootBlock = new Block(null, "2nd root block");
            var added = blockChain.AddBlock(rootBlock);

            Assert.IsFalse(added);
        }



        [TestMethod]
        public void TestAddBlock()
        {
            // nominal case for addBlock
            //TODO
            Assert.Fail();
        }


        [TestMethod]
        public void TestAddBlockWrongParent()
        {
            // adding a block with a parent not in the block chain
            //TODO
            Assert.Fail();
        }


        [TestMethod]
        public void TestGetLastBlock()
        {
            var rootBlock = blockChain.blocks["733ad20af52ec1902f51e97b99444754df7db2bf91bb26e076723c5e9fc12bc1"];
            var lastBlock = blockChain.GetLastBlock();

            Assert.AreSame(lastBlock, rootBlock); 
        }
        

        // Given this block chain:
        // a--b--c--d--e--f--g
        //     \-h--i
        // After adding a new block 'j' we should get:
        // a--b--c--d--e--f--g--j
        [TestMethod]
        public void TestCleanBlockChain()
        {
            var last = blockChain.GetLastBlock();
            var a = Add(last.hash, "a");
            var b = Add(a.hash, "b");
            var c = Add(b.hash, "c");
            var d = Add(c.hash, "d");
            var e = Add(d.hash, "e");
            var f = Add(e.hash, "f");
            var g = Add(f.hash, "g");
            var h = Add(b.hash, "h");
            var i = Add(h.hash, "i");

            var j = Add(g.hash, "j");

            CollectionAssert.DoesNotContain(blockChain.blocks.Keys, h.hash);
            CollectionAssert.DoesNotContain(blockChain.blocks.Keys, i.hash);
            CollectionAssert.Contains(blockChain.blocks.Keys, j.hash);
        }
    }
}
